meetings/working-groups/ref-improvements/REF-2022-11-11.md

```diff
@@ -0,0 +1,134 @@
+    where 'b <= 'a
+
+}
+```
+
+By establishing a relationship between `'a` and `'b` it allows for both the value and the `ref` to be return from the method. It also prevents cyclic assignment issues because the relationship does not allow for it from a lifetime (`'b` could be smaller than `'a`).
+
+# Complications around ref fields of ref struct
+The immediate issue this causes is it creates layers of lifetimes in our types. Consider concretely: 
+
+```csharp
+ref struct RS { 
+  ref 'a Span<byte> Field; 
+}
+```
+
+The value of the field has a lifetime that is independent of the containing instance. Typically when we look at a value there are two lifetimes to consider: the safe to escape and ref safe to escape. In the case of `RS` though there is also the lifetime of `Field`. 
+
+```csharp
+void E() {  
+    // Only two lifetimes here: the value and the ref 
+    Span<int> span = ...; 
+
+    // Three lifetimes: value, ref and value of Field
+    RS rs = ...;
+}
+```
+
+These lifetimes can now be arbitrarily deep. So a given struct can have `N` different lifetimes associated with it that we have to manage.  As the types get more fields this gets more complex because the lifetimes aren't necessarily linear either: 
+
+```csharp
+ref struct RS { 
+  ref Span<byte> Field1;
+  ref Span<byte> Field2;
+}
+``` 
+
+There are two different ways we could think about this: 
+
+```csharp
+// Option 1: simple but limiting
+ref struct RS { 
+  ref 'a Span<byte> Field1;
+  ref 'a Span<byte> Field2;
+}
+// Option 2: flexible but now we have a tree
+ref struct RS { 
+  ref 'a Span<byte> Field1;
+  ref 'b Span<byte> Field2;
+}
+```
+
+These problems mostly manifest in the method arguments must match rule. This is the rule where we look at all of the `ref` going into the method and ensure that they can never be cross assigned in a way that would cause an unsafe reference to the stack. 
+
+```csharp
+void Swap(ref Span<int> x, ref Span<int> y) {
+    ... 
+}
+
+void Use(ref Span<int> span) {
+    Span<int> local = stackalloc int[42];
+    Swap(ref span, ref local);
+}
+```
+The MAMM rules are intricate but can be simplified to the following: 
+
+> Every `ref` argument to a method be assignable *from* every other `ref` argument in terms of lifetimes. 
+
+Before there were just two layers of lifetimes, now there are basically infinite and they **all** have to match. The problem with `ref` fields to `ref struct` is now virtually everything is a MAMM nightmare. Even an instance method on a `struct` is dealing with MAMM insanity.
+
+The only way for MAMM to work in this scenario is for all the lifetimes involved to be equivalent. That really defeats the purpose of adding them. 
+
+# Paths Forward 
+
+## Implement RFRS
+The concensus is that to fully support RFRS would likely require us to have full lifetime notations in the langauge. For example putting the lifetimes on `scoped` modifier. That would allow for example `scoped<a>`, `scoped<b>`, etc ... 
```

Comment:
```suggestion
The consensus is that to fully support RFRS would likely require us to have full lifetime notations in the language. For example putting the lifetimes on `scoped` modifier. That would allow for example `scoped<a>`, `scoped<b>`, etc ... 
```